backend/x11/output.c

```diff
@@ -331,12 +338,179 @@ static void output_rollback_render(struct wlr_output *wlr_output) {
+	}
+	output->cursor.pic = XCB_NONE;
+
+	if (texture == NULL) {
+		return true;
+	}
+
+	if (output->cursor.swapchain == NULL ||
+			output->cursor.swapchain->width != width ||
+			output->cursor.swapchain->height != height) {
+		wlr_swapchain_destroy(output->cursor.swapchain);
+		output->cursor.swapchain = wlr_swapchain_create(
+			x11->allocator, width, height,
+			x11->drm_format);
+		if (output->cursor.swapchain == NULL) {
+			return false;
+		}
+	}
+
+	struct wlr_buffer *wlr_buffer =
+		wlr_swapchain_acquire(output->cursor.swapchain, NULL);
+	if (wlr_buffer == NULL) {
+		return false;
+	}
+
+	if (!wlr_renderer_bind_buffer(x11->renderer, wlr_buffer)) {
+		return false;
+	}
+
+	uint8_t *data = malloc(width * height * 4);
+	if (data == NULL) {
+		return false;
+	}
+
+	struct wlr_box cursor_box = {
+		.width = width,
+		.height = height,
+	};
+
+	float projection[9];
+	wlr_matrix_projection(projection, width, height, output->wlr_output.transform);
+
+	float matrix[9];
+	wlr_matrix_project_box(matrix, &cursor_box, transform, 0, projection);
+
+	wlr_renderer_begin(x11->renderer, width, height);
+	wlr_renderer_clear(x11->renderer, (float[]){ 0.0, 0.0, 0.0, 0.0 });
+	wlr_render_texture_with_matrix(x11->renderer, texture, matrix, 1.0);
+	wlr_renderer_end(x11->renderer);
+
+	bool result = wlr_renderer_read_pixels(
+		x11->renderer, WL_SHM_FORMAT_ARGB8888, NULL,
+		width * 4, width, height, 0, 0, 0, 0,
+		data);
+
+	wlr_renderer_bind_buffer(x11->renderer, NULL);
+
+	wlr_buffer_unlock(wlr_buffer);
+
+	if (!result) {
+		free(data);
+		return false;
+	}
+
+	xcb_pixmap_t pix = xcb_generate_id(x11->xcb);
+	xcb_create_pixmap(x11->xcb, depth, pix, output->win,
+		width, height);
+
+	output->cursor.pic = xcb_generate_id(x11->xcb);
+	xcb_render_create_picture(x11->xcb, output->cursor.pic,
+		pix, x11->argb32, 0, 0);
+
+	xcb_gcontext_t gc = xcb_generate_id(x11->xcb);
+	xcb_create_gc(x11->xcb, gc, pix, 0, NULL);
+
+	xcb_put_image(x11->xcb, XCB_IMAGE_FORMAT_Z_PIXMAP,
+		pix, gc, width, height, 0, 0, 0, depth,
+		stride * height * sizeof(uint8_t),
+		data);
+	free(data);
+	xcb_free_gc(x11->xcb, gc);
+	xcb_free_pixmap(x11->xcb, pix);
+
+	return true;
+}
+
+static bool output_set_cursor(struct wlr_output *wlr_output,
+		struct wlr_texture *texture, float scale,
+		enum wl_output_transform transform,
+		int32_t hotspot_x, int32_t hotspot_y, bool update_texture) {
+	struct wlr_x11_output *output = get_x11_output_from_output(wlr_output);
+	struct wlr_x11_backend *x11 = output->x11;
+	int width = 0, height = 0;
+
+	if (x11->argb32 == XCB_NONE) {
+		return false;
+	}
+
+	if (texture != NULL) {
+		width = texture->width * wlr_output->scale / scale;
+		height = texture->height * wlr_output->scale / scale;
+	}
+
+	struct wlr_box hotspot = { .x = hotspot_x, .y = hotspot_y };
+	wlr_box_transform(&hotspot, &hotspot,
+		wlr_output_transform_invert(wlr_output->transform),
+		width, height);
+
+	// TODO: use output->wlr_output.transform to transform pixels and hotpot
```

Comment:
This seems implemented.

Reply:
OK. That was just some copy/paste from the wayland one.